File: python/29_Divide_Two_Integers.py
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        i,a,b = 0,abs(dividend),abs(divisor)
        if a==0 or a<b:
            return 0
        while b <= a:
            b = b << 1
            i = i + 1
        else:
            res = (1 << (i-1)) + self.divide(a-(b>>1),abs(divisor))
            if(dividend ^ divisor) < 0:
                res *= -1
            return min(res,(1<<31) -1)

def main():
    import sys
    import io
    def readlines():
        for line in io.TextIOWrapper(sys.stdin.buffer, encoding='utf-8'):
            yield line.strip('\n')

    lines = readlines()
    while True:
        try:
            line = next(lines)
            dividend = int(line);
            line = next(lines)
            divisor = int(line);
            
            ret = Solution().divide(dividend, divisor)

            out = str(ret);
            print(out)
        except StopIteration:
            break

if __name__ == '__main__':
    main()

File: java/29_Divide_Two_Integers.java
class Solution {
    public int divide(int dividend, int divisor) {
        if( dividend == 0){
            return 0;
        }
        if(dividend == Integer.MIN_VALUE && divisor == -1){
            return Integer.MAX_VALUE;
        }
        boolean negative = (dividend ^ divisor) < 0;
        long t = Math.abs((long)dividend);
        long d = Math.abs((long)divisor);
        int result = 0 ;
        for(int i = 31 ; i >= 0 ; i--){
            if((t>>i) >= d){
                result += 1<<i;
                t -= d<<i;
            }
        }
        return negative ? -result : result;
    }
}

public class MainClass {
    public static void main(String[] args) throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        String line;
        while ((line = in.readLine()) != null) {
            int dividend = Integer.parseInt(line);
            line = in.readLine();
            int divisor = Integer.parseInt(line);
            
            int ret = new Solution().divide(dividend, divisor);
            
            String out = String.valueOf(ret);
            
            System.out.print(out);
        }
    }
}